src/source/Ice/IceAgent.c

```diff
@@ -614,107 +620,105 @@ STATUS createIceCandidatePairs(PIceAgent pIceAgent, PIceCandidate pIceCandidate,
-    UINT64 pivotPriority;
-    BOOL auxStackEmpty = FALSE;
-    PIceCandidatePair temp;
+    PIceCandidatePair pIceCandidatePair = NULL;
 
-    CHK(pIceAgent != NULL, STATUS_NULL_ARG);
-    CHK(pIceAgent->candidatePairCount != 0, retStatus);
-
-    low = 0;
-    high = pIceAgent->candidatePairCount - 1;
-    CHK_STATUS(stackQueueCreate(&pAuxStack));
-    CHK_STATUS(stackQueuePush(pAuxStack, low));
-    CHK_STATUS(stackQueuePush(pAuxStack, high));
-
-    while(!auxStackEmpty) {
-        CHK_STATUS(stackQueuePop(pAuxStack, &high));
-        CHK_STATUS(stackQueuePop(pAuxStack, &low));
-
-        // partition in descending order. pivot is swapped at the last iteration.
-        pivotPriority = pIceAgent->candidatePairs[high]->priority;
-        for (j = low, i = low; j <= high; ++j) {
-            if (pIceAgent->candidatePairs[j]->priority >= pivotPriority) {
-                temp = pIceAgent->candidatePairs[j];
-                pIceAgent->candidatePairs[j] = pIceAgent->candidatePairs[i];
-                pIceAgent->candidatePairs[i] = temp;
-                i++;
-            }
-        }
+    CHK(ppIceCandidatePair != NULL, STATUS_NULL_ARG);
+    // free is idempotent
+    CHK(*ppIceCandidatePair != NULL, retStatus);
+    pIceCandidatePair = *ppIceCandidatePair;
 
-        // set i to index of pivot
-        i--;
+    CHK_LOG_ERR_NV(freeTransactionIdStore(&pIceCandidatePair->pTransactionIdStore));
+    SAFE_MEMFREE(pIceCandidatePair);
 
-        // sort values before pivot
-        if (i > 1) {
-            CHK_STATUS(stackQueuePush(pAuxStack, 0));
-            CHK_STATUS(stackQueuePush(pAuxStack, i - 1));
-        }
+CleanUp:
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS insertIceCandidatePair(PDoubleList iceCandidatePairs, PIceCandidatePair pIceCandidatePair)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    PDoubleListNode pCurNode = NULL;
+    PIceCandidatePair pCurIceCandidatePair = NULL;
+    UINT32 currIceCandidateCount = 0;
+
+    CHK(iceCandidatePairs != NULL && pIceCandidatePair != NULL, STATUS_NULL_ARG);
 
-        // sort values after pivot
-        if (high - i > 1) {
-            CHK_STATUS(stackQueuePush(pAuxStack, i + 1));
-            CHK_STATUS(stackQueuePush(pAuxStack, high));
+    CHK_STATUS(doubleListGetHeadNode(iceCandidatePairs, &pCurNode));
+
+    while(pCurNode != NULL) {
+        pCurIceCandidatePair = (PIceCandidatePair) pCurNode->data;
+
+        // insert new candidate pair ordered by priority from max to min.
+        if (pCurIceCandidatePair->priority <= pIceCandidatePair->priority) {
+            break;
         }
+        pCurNode = pCurNode->pNext;
+    }
+
+    CHK_STATUS(doubleListGetNodeCount(iceCandidatePairs, &currIceCandidateCount));
 
-        CHK_STATUS(stackQueueIsEmpty(pAuxStack, &auxStackEmpty));
+    if (pCurNode != NULL) {
+        CHK_STATUS(doubleListInsertItemBefore(iceCandidatePairs, pCurNode, (UINT64) pIceCandidatePair));
+    } else if (currIceCandidateCount == 0) {
```

Comment:
I think when pCurNode == NULL you can always just doubleListInsertItemTail.

Reply:
fixed